### PR TITLE
Read mitgrid files

### LINKED ISSUE #14: read mitgrid files
It would be good to read the ``tile00?.mitgrid`` files directly in order to load the proper grid data. Many of the LLC files have blank tiles in the XC.data etc. files.

### PR DESCRIPTION
**Pull request by raphaeldussin**
fixes #14 

tested with llc90 and aste. there is room to add cube sphere support once #91 is resolved
and it would be also easy to make it work with simpler geometries, although I have not seen
yet the case of users using input grid files in those.

### CODE REVIEW

**Comment by rabernat** [comment]
This looks very useful!

Tests and documentation are necessary.

File: xmitgcm/utils.py [hunk 1]
```diff
@@ -1125,3 +1126,153 @@ def get_extra_metadata(domain='llc', nx=90):
...
+                       'DXV': (['face', 'j', 'i'],     gridfields['DXV']),
+                       'DYU': (['face', 'j', 'i'],     gridfields['DYU']),
+                       'RAZ': (['face', 'j_g', 'i_g'], gridfields['RAZ']),
+                       'DXC': (['face', 'j', 'i_g'],   gridfields['DXC']),
+                       'DYC': (['face', 'j_g', 'i'],   gridfields['DYC']),
+                       'RAW': (['face', 'j', 'i_g'],   gridfields['RAW']),
+                       'RAS': (['face', 'j_g', 'i'],   gridfields['RAS']),
+                       'DXG': (['face', 'j_g', 'i'],   gridfields['DXG']),
+                       'DYG': (['face', 'j', 'i_g'],   gridfields['DYG'])
+                       })
```
Comment (rabernat) [inline xmitgcm/utils.py]: This dataset is missing the coordinates for `j` and `i`. In a regular xmitgcm dataset, these variables have explicit coordinates.

File: xmitgcm/utils.py [hunk 2]
```diff
@@ -1125,3 +1126,153 @@ def get_extra_metadata(domain='llc', nx=90):
...
+
+            rawfields = {}
+            for kfield in np.arange(len(file_metadata['fldList'])):
+
+                rawfields.update(
+                    {file_metadata['fldList'][kfield]: raw[kfield]})
+
+            for field in file_metadata['fldList']:
+                # symetrize
+                tmp = rawfields[field][:, :, :-1, :-1].squeeze()
```
[the remark this answers is not in the document]
Comment (raphaeldussin) [reply xmitgcm/utils.py]: it's all zeros so no worries.

```python
import numpy as np
all = np.fromfile('tile003.mitgrid','>d')
XC = all[:91*91]
XC = np.reshape(XC,(91,91))
```

In [14]: XC[-1,:]
Out[14]:
array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0.])

In [15]: XC[:,-1]
Out[15]:
array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0.])

File: xmitgcm/utils.py [hunk 3]
```diff
@@ -1125,3 +1126,153 @@ def get_extra_metadata(domain='llc', nx=90):
...
+                        dataface = tmp[offset*nx:(offset+1)*nx, :]
+                        # transpose, if needed
+                        if file_metadata['transpose_face'][face]:
+                            dataface = dataface.transpose()
+                        # assign values
+                        gridfields[field][face, :, :] = dataface
+
+    elif geometry == 'cs':
+        # TO DO
+        pass
```
[the remark this answers is not in the document]
Comment (raphaeldussin) [reply xmitgcm/utils.py]: done

File: xmitgcm/utils.py (hunk 1 above)
Comment (raphaeldussin) [reply xmitgcm/utils.py]: done

File: xmitgcm/utils.py [hunk 4]
```diff
@@ -1125,3 +1126,153 @@ def get_extra_metadata(domain='llc', nx=90):
...
+                        nx = file_metadata['nx']
+                        # pad data, if needed
+                        tmp = _pad_array(tmp, file_metadata, face=face)
+                        # extract the data
+                        dataface = tmp[offset*nx:(offset+1)*nx, :]
+                        # transpose, if needed
+                        if file_metadata['transpose_face'][face]:
+                            dataface = dataface.transpose()
+                        # assign values
+                        gridfields[field][face, :, :] = dataface
```
[the remark this answers is not in the document]
Comment (raphaeldussin) [reply xmitgcm/utils.py]: I haven't seen anywhere in xmitgcm functions working on multi-files but I can have missed it.

File: xmitgcm/utils.py [hunk 5]
```diff
@@ -1125,3 +1126,153 @@ def get_extra_metadata(domain='llc', nx=90):
...
+                nxgrid = file_metadata['nx'] + 1
+                nygrid = file_metadata['ny_facets'][kfacet] + 1
+            elif file_metadata['facet_orders'][kfacet] == 'F':
+                nxgrid = file_metadata['ny_facets'][kfacet] + 1
+                nygrid = file_metadata['nx'] + 1
+
+            grid_metadata.update({'nx': nxgrid, 'ny': nygrid,
+                                  'has_faces': False})
+
+            raw = read_all_variables(grid_metadata['vars'], grid_metadata)
```
[the remark this answers is not in the document]
Comment (raphaeldussin) [reply xmitgcm/utils.py]: done

**Comment by raphaeldussin** [comment]
@rabernat strickler-ci is really nitpicking, can we merge this anyway? I will fill in the blanks with the cube sphere grid in the cs branch. Not very orthogonal here I'm sorry!

File: xmitgcm/test/test_utils.py [hunk 6]
```diff
@@ -843,3 +843,15 @@ def test_get_extra_metadata(domain, nx):
...
+    from xmitgcm.utils import get_grid_from_input, get_extra_metadata
+    dirname, expected = all_grid_datadirs
+    md = get_extra_metadata(domain=expected['domain'], nx=expected['nx'])
+    ds = get_grid_from_input(dirname + '/' + expected['gridfile'], 
+                             geometry=expected['geometry'],
+                             precision='double', endian='>', 
+                             use_dask=False,
+                             extra_metadata=md)
+    assert type(ds) == xarray.Dataset
+    assert ds['XC'].values.shape == expected['shape']
```
[the remark this answers is not in the document]
Comment (rabernat) [reply xmitgcm/test/test_utils.py]: maybe add a check here that all the expected variables are in `ds`?

File: xmitgcm/utils.py [hunk 7]
```diff
@@ -1125,3 +1126,193 @@ def get_extra_metadata(domain='llc', nx=90):
...
+    if geometry == 'cs':  # pragma: no cover
+        raise NotImplementedError("'cs' geometry is not supported yet")
+
+    # create placeholders for data
+    gridfields = {}
+    for field in file_metadata['fldList']:
+        gridfields.update({field: np.zeros(shape)})
+
+    if geometry == 'llc':
+        for kfacet in np.arange(nfacets):
```
Comment (rabernat) [inline xmitgcm/utils.py]: `range(nfacets)` would be better here

File: xmitgcm/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rabernat) [reply xmitgcm/utils.py]: Does this work with dask?

File: xmitgcm/test/test_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rabernat) [reply xmitgcm/test/test_utils.py]: Also need to check that `use_dask` works as expected and doesn't trigger eagerreading of the data.

**Comment by raphaeldussin** [comment]
@rabernat dask should work with global llc grids (not regional because of the padding), the dsa.concatenate on l1260 in utils.py makes the results being a dask array wether the input is a dask or numpy array so I can't test for the type in the output.

**Comment by rabernat** [comment]
> dsa.concatenate on l1260 in utils.py makes the results being a dask array wether the input is a dask or numpy array so I can't test for the type in the output.

Ok, then maybe we can come back to this in a future PR. I would prefer to use a version of concatenate that returns numpy arrays for numpy inputs. We have something like that in xgcm:
https://github.com/xgcm/xgcm/blob/master/xgcm/duck_array_ops.py#L37

**Pull request merged**